vibhaa/lnd
# Stats

All the statistics are logged in htlcswitch/link.go/ in the function: updateFirebaseDB().

This PR is built on top of the PR at: https://github.com/vibhaa/lnd/pull/2/  But it fixes the issues in that PR when the queue functionality is not being used - and appears to be definitely correct when SPIDER_QUEUE is not used, while I am still working on verifying / fixing the behavior when the QUEUE is turned on. 

The behavior for the firebase logging is controlled by environment variables.
tldr; for running it locally, do the following:
	export SPIDER_LOG_FIREBASE=1
	export SPIDER_QUEUE=0
	export SPIDER_EXP_NAME="new name"

More details: 

	- SPIDER_QUEUE: 1 means queue will be used, otherwise the queue related changes will not be used
	- SPIDER_LOG_FIREBASE: 1 means Spider will log information every second to firebase, otherwise it will not.  (Note: The frequency of the updates can be changed by using the variable  in htlcswitch/spider_utils.go)

	- SPIDER_EXP_NAME: Only used for the purpose of logging to firebase, this defines a unique name for the experimental run. 

	- NODENAME: If specified, this is used as the unique name for the switch running in that environment. It only makes sense in a distributed setting, as then each node would have it's own dockerized environment. In the local setting, this should be left unspecified, or "". Each switch will then be identified by the hash of it's configuration (this is unique and stays constant through the run). In the visualizations, we see only the first 3 characters of the hash for simplicity.

After setting these environment variables, you can test that everything is working with these. For a short test, from the /htlcswitch directory, run:
	-  go test -run TestSpiderTemporarilyInsufficientFundsMultiple
Or from the root lnd directory:
	- make itest

Once you have run the experiment, you can verify that the results were collected at: https://console.firebase.google.com/u/0/project/lnd-test-1dd52/database/lnd-test-1dd52/data/.

To run the visualizations, the code is at: https://github.com/vibhaa/lightning_routing/tree/master/lnd-vis/

Instructions:
	- install visdom: pip install visdom
	- start a visdom server from the command line: visdom -p PORT
	- python main.py -exp_name <whatever-name-you-specified-before>
	- e.g., I have some visualizations running in `chunky', which can be accessed at: http://chunky.csail.mit.edu:4000/
